core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/job/report/DefaultReportJobDelegate.java

```diff
@@ -124,6 +126,8 @@ public void execute(final String reportKey) throws JobExecutionException {
         execution.setStatus(ReportExecStatus.STARTED);
         execution.setStart(new Date());
         execution.setReport(report);
+
+        execution.setExecutor(AuthContextUtils.getUsername());
```

Comment:
Report execution can be either triggered [via REST](https://github.com/apache/syncope/blob/master/core/idrepo/logic/src/main/java/org/apache/syncope/core/logic/ReportLogic.java#L163) or scheduled; either way, it will be submitted to Quartz.

[ReportJob](https://github.com/apache/syncope/blob/master/core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/job/report/ReportJob.java) is executed by Quartz, and in the block https://github.com/apache/syncope/blob/master/core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/job/report/ReportJob.java#L64-L74 will invoke the configured `ReportJobDelegate` instance, which is by default this class, e.g. `DefaultReportJobDelegate`.

As you can see, an instrumented authentication context is built there for user `admin`; as a result, `AuthContextUtils.getUsername()` will always return `admin` here, which, at least for executions triggered via REST, may not be correct.

In order to get the actual executor, you should invoke `AuthContextUtils.getUsername()` in `ReportLogic#execute`, then pass such value to `jobManager#register` and add it in the Map that will be passed inside `JobExecutionContext` to `ReportJob#execute` an replace there `AuthContextUtils#callAsAdmin` with `AuthContextUtils#callAs`.

Naturally, you should also adjust things so that `admin` is used when execution trigger is not coming via REST but scheduled.

Reply:
Forgot to add: it would be more or less the same for Tasks.

Reply:
Thanks very much for the notes; the design makes much better sense now. Will adjust.

Reply:
After thinking again to the change I propose above, I believe we should keep calling `AuthContextUtils#callAsAdmin` and add `executor` as String param to `ReportJobDelegate#execute` - it's simpler and also more in line with similar use cases as audit and notifications.

Reply:
Question:

If I modify the `JobManagerImpl` to be:

```
public Map<String, Object> register(final SchedTask task, final Date startAt, 
									final long interruptMaxRetries,
                                    final String executor) {

	...
	jobMap.put(JobManager.EXECUTOR_KEY, executor);
    ...
}
```


That would mean that I also may need to modify:

```
public void load(final String domain, final DataSource datasource) {}
```

Because, it calls the `register` function [here](https://github.com/apache/syncope/blob/master/core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/job/JobManagerImpl.java#L353) in `load()`, and this would cause API changes to many other components.

So, in `JobManagerImpl#load()`, we could just call `AuthContextUtils.getUsername()` such that

```
register(task, task.getStartAt(), conf.getRight(), AuthContextUtils.getUsername());
```

Would that work?

Reply:
> Would that work?

Nope. 

`JobManagerImpl#load()` is invoked at startup: at that point, no authentication context exists, of course.

Hence, you can add

```java
    @Resource(name = "adminUser")
    private String adminUser;
```

to `JobManagerImpl` and then invoke

```java
register(task, task.getStartAt(), conf.getRight(), adminUser);
```

Incidentally, this is exactly what I mean above by saying that report and task execution can be scheduled.